framework/mongo/__init__.py

```diff
@@ -1,10 +1,41 @@
 # -*- coding: utf-8 -*-
 
-from modularodm import FlaskStoredObject as StoredObject
+from flask import request
+from modularodm.storedobject import StoredObject as GenericStoredObject
+from modularodm.ext.concurrency import with_proxies, proxied_members
 
 from bson import ObjectId
 from .handlers import client, database, set_up_storage
 
+
+from api.base.api_globals import api_globals
+
+
+class DummyRequest(object):
+    pass
+dummy_request = DummyRequest()
+
+
+def get_cache_key():
+    """
+    Fetch a request key from either a Django or Flask request. Fall back on a process-global dummy object
+    if we are not in either type of request
+    """
+    # TODO: This is ugly use of exceptions; is there a better way to track whether in a given type of request?
+    try:
+        return request._get_current_object()
+    except RuntimeError:  # Not in a flask request context
+        if hasattr(api_globals, 'request') and api_globals.request is not None:
+            return api_globals.request
+        else:  # Not in a Django request
+            return dummy_request
```

Comment:
When changed to use a new instance of dummy request in every fallback cache request...
```diff
diff --git a/framework/mongo/__init__.py b/framework/mongo/__init__.py
index 2ff142b..4534b07 100644
--- a/framework/mongo/__init__.py
+++ b/framework/mongo/__init__.py
@@ -28,7 +28,7 @@ def get_cache_key():
         if hasattr(api_globals, 'request') and api_globals.request is not None:
             return api_globals.request
         else:  # Not in a Django request
-            return dummy_request
+            return DummyRequest()
```
The following error occurs when running invoke_server. It can be traced to calling ensure_schemas within the call to init_app.

```python
[root]  INFO: Sentry disabled; Flask's debug mode enabled
Traceback (most recent call last):
  File "/Users/andyboughton/code/cos/osf.io/env/bin/inv", line 9, in <module>
    load_entry_point('invoke==0.9.0', 'console_scripts', 'inv')()
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/invoke/cli.py", line 304, in main
    dispatch(sys.argv)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/invoke/cli.py", line 297, in dispatch
    return executor.execute(*tasks, dedupe=dedupe)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/invoke/executor.py", line 90, in execute
    task=task, name=name, args=args, kwargs=kwargs
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/invoke/executor.py", line 131, in _execute
    result = task(*args, **kwargs)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/invoke/tasks.py", line 111, in __call__
    result = self.body(*args, **kwargs)
  File "/Users/andyboughton/code/cos/osf.io/tasks.py", line 47, in server
    app = init_app(set_backends=True, routes=True)
  File "/Users/andyboughton/code/cos/osf.io/website/app.py", line 135, in init_app
    ensure_schemas()
  File "/Users/andyboughton/code/cos/osf.io/website/project/model.py", line 103, in ensure_schemas
    MetaSchema.remove()
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 110, in wrapped
    return func(*args, **kwargs)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 1231, in remove
    cls.remove_one(obj, rm=False)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 110, in wrapped
    return func(*args, **kwargs)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 1207, in remove_one
    cls._clear_caches(obj._storage_key)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 500, in _clear_caches
    cls._clear_data_cache(key)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/storedobject.py", line 485, in _clear_data_cache
    cls._cache.pop(cls._name, key)
  File "/Users/andyboughton/code/cos/osf.io/env/lib/python2.7/site-packages/modularodm/cache.py", line 45, in pop
    self.data[schema].pop(key, None)
KeyError: 'metaschema'
```